src/Compilers/Core/Portable/DiagnosticAnalyzer/SuppressMessageAttributeState.cs

```diff
@@ -213,19 +246,34 @@ private ImmutableArray<string> DecodeSuppressMessageAttributes(ISymbol symbol)
                     continue;
                 }
 
-                builder.Add(info.Id);
+                AddOrUpdate(info, builder);
             }
 
-            return builder.ToImmutableAndFree();
+            return builder.ToImmutable();
+        }
+
+        private static void AddOrUpdate(SuppressMessageInfo info, IDictionary<string, SuppressMessageInfo> builder)
+        {
+            // TODO: How should we deal with multiple SuppressMessage attributes, with different suppression info/states?
+            // For now, we just pick the last attribute, if not suppressed.
+            SuppressMessageInfo currentInfo;
+            if (!builder.TryGetValue(info.Id, out currentInfo))
```

Comment:
Looks like all you are doing is checking whether the key is present? If so, can't you just use IDictionary.ContainsKey()?

Also looks like the function will never really 'update' - it will only 'add' (if the key doesn't already exist).Perhaps just rename the function from 'AddOrUpdate' to 'Add'?

Finally, the TODO comment seems stale. Looks like currently in the event of multiple attributes, we will simply pick the 'first' one as opposed to the 'last' one as the comment seems to imply.